[RemoteMirror] Get spare bit info from reflection records

tbkka opened the pull request:
This adds a new reflection record type carrying spare bit information for multi-payload enums.

The compiler includes this for any type that might need it in order to accurately reflect the contents of the enum.  The RemoteMirror library will use this if present to determine how to project the contents of the enum.  If not present (for example, in older binaries), the RemoteMirror library falls back on an internal calculation of the spare bitmask.

Resolves rdar://61158214

tbkka committed:
Initialize FieldInfo

tbkka committed:
Expanded tests for MPEs

tbkka committed:
Restore fallback calculation of MPE spare bit mask

tbkka committed:
Add MPE flag bits to the BuiltinTypeDescriptor

tbkka committed:
Store MPE layout in a new MultiPayloadEnumDescriptor reflection record

I'm still not getting the correct bitmask data into this record,
but now I've at least managed to both store real data into
the object file and recover it from inside the reflection library.

tbkka commented on stdlib/public/Reflection/TypeRefBuilder.cpp:
```diff
@@ -496,6 +525,15 @@ void TypeRefBuilder::dumpCaptureSection(std::ostream &stream) {
+  for (const auto &sections : ReflectionInfos) {
+    for (const auto descriptor : sections.MultiPayloadEnum) {
+//      auto info = getXXXXX(descriptor);
+//      info.dump(stream);
```
TODO

tbkka replied:
I've filled in the next layer of `dump` functions.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -1890,33 +1999,70 @@ class EnumTypeInfoBuilder {
+              // TEMPORARY: compare our computed spare bit mask to the
+              // one we got from the compiler (to verify that the compiler
+              // is putting the right data in)
+              assert(t == spareBitsMask);
```
TODO: Remove this assertion once things are more fully tested.

tbkka replied:
I've converted this to only be included when NDEBUG is false.

tbkka commented on lib/IRGen/GenReflection.cpp:
```diff
@@ -1006,6 +1006,65 @@ void IRGenModule::emitBuiltinTypeMetadataRecord(CanType builtinType) {
+
+    addTypeRef(type, CanGenericSignature());
+
+    auto &strategy = getEnumImplStrategy(IGM, type);
```
TODO: getEnumImplStrategy here is crashing, apparently because `type` is sometimes an unbound generic?  I don't believe I need to emit an MPE descriptor for unbound generics, so it should suffice to omit the MPE descriptor in that case.

tbkka replied:
@aschwaighofer Can you help me figure out what `type` I should be using here?

aschwaighofer replied:
You will want something like `CanType getFormalTypeInContext(CanType abstractType)` -- it maps the interface types into the context of the type's generic signature.

tbkka replied:
I copied that function from `GenEnum.cpp` and it seems to stop the crashing.
TODO: Share that function instead of copying it.

tbkka replied:
TODO copied below; rest of this seems to be working.

tbkka commented on lib/IRGen/GenReflection.cpp:
```diff
@@ -1451,6 +1517,7 @@ void IRGenModule::emitFieldDescriptor(const NominalTypeDecl *D) {
     if (strategy.getElementsWithPayload().size() > 1 &&
         !strategy.needsPayloadSizeInMetadata()) {
       needsOpaqueDescriptor = true;
+      needsMPEDescriptor = true;
```
TODO?  Is this the right place to suppress MPE descriptors for unbound generic enums?

tbkka replied:
After some more inspection, I think this is actually correct already.

tbkka commented on include/swift/SwiftRemoteMirror/SwiftRemoteMirrorTypes.h:
```diff
@@ -77,6 +77,7 @@ typedef struct swift_reflection_info {
   swift_reflection_section_pair_t capture;
   swift_reflection_section_pair_t type_references;
   swift_reflection_section_pair_t reflection_strings;
+  swift_reflection_section_pair_t multipayload_enums; // TODO: Is this okay?
```
Q: Is it okay to just add another field here?  Does that mess up existing clients?  Is there an alternative?

tbkka replied:
@mikeash Do you know what the consequences here are of adding another field for a new section?

tbkka replied:
Thanks for clarifying.  I've backed out this change.

tbkka commented on lib/IRGen/GenReflection.cpp:
```diff
@@ -1006,6 +1006,65 @@ void IRGenModule::emitBuiltinTypeMetadataRecord(CanType builtinType) {
+    const TypeInfo &TI = strategy.getTypeInfo();
+    auto bits1 = strategy.getTagBitsForPayloads();
+    auto bits2 = strategy.getBitMaskForNoPayloadElements();
+    // Look at EnumImplStrategy::emitGetEnumTag to see how it works
```
TODO:  Get the spare bitmask from the right place.  `bits1` and `bits2` above look like either one of them might be right.

tbkka replied:
Looking at [GenEnum.cpp:5572](https://github.com/apple/swift/blob/a193d2d791f9ae972c1689318f3dc20cbc7be751/lib/IRGen/GenEnum.cpp#L5572), it really looks like `getTagBitsForPayloads()` should provide exactly what I need here.  But for an MPE where every payload is a class reference, it is giving me 0xe000000000000000 instead of the 0xff00000000000007 that I expect.  @slavapestov What am I missing here?

aschwaighofer replied:
Looks like you might have an enum in your test case that requires 3bits to represent the tag? `getTagBitsForPayloads()` returns the bits that are used for tagging. Not the spare bits available.

```
    // The common spare bits actually used for a tag in the payload area.       
    SpareBitVector PayloadTagBits;  
```

`IRGenModule::getHeapObjectSpareBits` would tell you what spare bits we use for heap objects.

Assuming you are looking for the common spare bits among the payloads of an multi payload enum (?) I think you want (to add) a function that returns:

```
    // The spare bits shared by all payloads, if any.                           
    // Invariant: The size of the bit vector is the size of the payload in bits,
    // rounded up to a byte boundary.                                           
    SpareBitVector CommonSpareBits;     
```

tbkka replied:
After a few hours of tracing code, I found [`applyFixedSpareBitsMask`](https://github.com/apple/swift/blob/a193d2d791f9ae972c1689318f3dc20cbc7be751/lib/IRGen/FixedTypeInfo.h#L219), which seems to do exactly that.

aschwaighofer replied:
I am not sure what mask you are looking for. 

But I don't believe that `multipayloadEnum->applyFixedSpareBitsMask() == CommonSpareBits`. I would expect `multipayloadEnum->applyFixedSpareBitsMask()` to return the remaining spare bits after subtracting the bit used for the tag (also see the code at https://github.com/apple/swift/blob/877fb2b84110d88c11c39e98bd9b3ac292f9c284/lib/IRGen/GenEnum.cpp#L6642)

A useful tools debugging inspecting this is to use the command:

```
swiftc -Xllvm -debug-only=enum-layout fileWithEnums.swift -c
```

This will enable the debug printing code at https://github.com/apple/swift/blob/877fb2b84110d88c11c39e98bd9b3ac292f9c284/lib/IRGen/GenEnum.cpp#L6901

If you add code to print the `CommonSpareBits` you can view how they relate to the other values. On first appearance I would think that you can reconstruct CommonSpareBits by combining `applyFixedSpareBitsMask` and `getTagBitsForPayloads` (assuming we don't use extra tag bits at least?).

tbkka replied:
The right code here turns out to be the following, which computes the common spare bits for all payloads (which is precisely what the runtime is looking for):
```
      for (auto enumCase : strategy.getElementsWithPayload()) {
        cast<FixedTypeInfo>(enumCase.ti)->applyFixedSpareBitsMask(IGM, spareBits);
      }
```

tbkka commented on include/swift/Reflection/Records.h:
```diff
@@ -374,6 +374,22 @@ class BuiltinTypeDescriptor {
+  const RelativeDirectPointer<const char> TypeName;
+  uint32_t Flags;
+  uint32_t BitfieldBitCount; // Size of bitfield in bits
+  uint8_t  BitfieldBits[]; // Variably-sized bitmask field; always a multiple of 4 bytes
```
TODO: This descriptor is inherently variably-sized.  Should I include an explicit `size` field so it can potentially be expanded in the future?

tbkka replied:
@mikeash Should I worry about evolution for reflection records?  It doesn't look like our existing reflection records have any support for being expanded in the future.  This seems potentially problematic, though maybe it's okay to "just" add a new section every time we discover we need a new field?

tbkka replied:
I worked up an architecture here that allows this record to add new properties in the future:  There's a single `uint32_t contents[]` array that holds all data, and standard boilerplate to compute the index of any particular field stored in that array.  I've included a commented-out template showing how to add a new property.

al45tair commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -703,21 +745,31 @@ class BitMask {
   void andMask(IntegerType value, unsigned byteOffset) {
-    andMask((void *)&value, sizeof(value), byteOffset);
+    if (byteOffset >= size) {
+      makeZero();
```
This seems to imply that `andMask()` does a bitwise and of the `BitMask` with `andMask()`'s argument, treating the latter as zero extended in either direction. However, the implementation of `andMask()` below doesn't seem to match that behaviour — it doesn't zero bytes in the `BitMask` that are not covered by the `andMask()`'s argument.

tbkka replied:
Thanks for pointing that out.  It actually looks like the uses of it below expect that the argument is one-extended in either direction.  I'll comment that expectation and fix up the implementation here to be consistent.

al45tair commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -789,17 +841,20 @@ class BitMask {
+    for (unsigned i = 0; i < common; ++i) {
       mask[i + offset] &= maskBytes[i];
     }
   }
```
Should this zero `mask[i]` from `i=0` to `i=offset-1` and from `i=common+offset` to `i=size`? That would seem consistent with the behaviour of the `andMask()` functions above.

al45tair commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -987,6 +1042,60 @@ class MultiPayloadEnumTypeInfo: public EnumTypeInfo {
+    // Or else somehow get access to target architecture information?
+    if (TI->getSize() == 8) {
+      // x86_64 masking (ARM64 is different)
+      mask.andMask((uint64_t)0xFF00000000000007ULL, 0);
```
There must surely be constants for these somewhere? Or if not, we should add them?

tbkka replied:
They're in stdlib/public/SwiftShims/System.h.  Using them is a bit tricky.  The reflection library here doesn't actually know much about the target other than the pointer size:  right now, we use the same mask for all 32-bit platforms, but not for all 64-bit, so I think the best I can do here is assume that if the pointer size is the same as the host, then it's the same architecture.  (We actually assume that a lot of places in the reflection code.)

tbkka replied:
I've improved this:  If the target pointer size is the same as ours, assume the target is the same architecture as us.  Otherwise, assume ARM64 for 8-byte pointer target and ARM32 for 4-byte pointer targets.  The last here is fine, as all 32-bit targets currently use the same mask.  To do better for 8-byte pointer targets, we'd need some way to determine the CPU architecture of the target which I don't believe we have yet.

mikeash replied:
We've taken a first step toward differentiating between targets in ways other than pointer size, with the ability to specify whether the target does ObjC interop. See: https://github.com/apple/swift/pull/40927/files. The right fix here would be to add another thing for x86 or ARM64 style masking. We can, of course, figure that out later.

tbkka replied:
Would it make more sense to expose the target client triple to the host?  I guess that won't necessarily work if things like the mask style might change over time for a particular target.  Hmmm....

mikeash replied:
Another option would be to place the mask in a variable in `Debug.h`, then read the value out of it here. As long as the runtime knows the correct mask when it's built, that would handle all scenarios.

tbkka replied:
Hmmm....  There are four different `Debug.h` files in the source tree.  I presume you mean `include/swift/Runtime/Debug.h`

mikeash replied:
Yeah, that's the one. We could add another `_swift_debug` variable that contains the mask.``

tbkka replied:
The only place where I seem to be able to actually inspect the other process is from `TypeRefBuilder`.  But I don't understand the lifetime of a `TypeRefBuilder`.  It's initialized with a _reader_ object that holds a `Reader` which can actually inspect the other process.  But that `Reader` is just nonsense at initialization time -- sometimes it's NULL, sometimes it's just garbage.  So presumably `Reader` is filled in sometime later?

mikeash replied:
The constructor for`ReflectionContext` takes a `MemoryReader` and passes it along to `super`:
```
  explicit ReflectionContext(std::shared_ptr<MemoryReader> reader)
    : super(std::move(reader), *this)
  {}
```
`super` is `MetadataReader<Runtime, TypeRefBuilder>`. Its constructor creates a `TypeRefBuilder` with the reader:
```
  template <class... T>
  MetadataReader(std::shared_ptr<MemoryReader> reader, T &&... args)
    : Builder(std::forward<T>(args)...),
      Reader(std::move(reader)),
      PtrAuthMask(queryPtrAuthMask()) {
  }
```
So everything should be set up once `TypeRefBuilder` is constructed. Not sure what would be going on there.

tbkka replied:
The tests that are failing for me are the ones based on `swift-reflection-dump`, such as `test/Reflection/typeref_lowering.swift`

tbkka replied:
Specifically, applying the following patch to `TypeRefBuilder.h` causes `test/Reflection/typeref_lowering.swift` to pretty reliably blow up:
```
--- a/include/swift/Reflection/TypeRefBuilder.h
+++ b/include/swift/Reflection/TypeRefBuilder.h
@@ -774,7 +774,36 @@ public:
       OpaquePointerReader([&reader](remote::RemoteAddress address, unsigned size) -> llvm::Optional<remote::RemoteAbsolutePointer> {
         return reader.Reader->readPointer(address, size);
       })
-  {}
+  {
+    if (auto Reader = reader.Reader) {
+      std::string TestName = "_swift_debug_foo";
+      auto TestAddr = Reader->getSymbolAddress(TestName);
+      uint64_t mask = 0;
+      if (TestAddr) {
+        std::cerr << "Found " << TestName << std::endl;
+      } else {
+        std::cerr << "Failed to find address of " << TestName << std::endl;
+      }
+    } else {
+      std::cerr << "No Reader?!?!" << std::endl;
+    }
+  }
 
   Demangle::Node *demangleTypeRef(RemoteRef<char> string,
                                   bool useOpaqueTypeSymbolicReferences = true) {
```

mikeash replied:
Looks like that test uses `include/swift/StaticMirror/ObjectFileContext.h` which implements `getSymbolAddress` to always return `NULL`:
```
  // TODO: We could consult the dynamic symbol tables of the images to
  // implement this.
  reflection::RemoteAddress getSymbolAddress(const std::string &name) override {
    return reflection::RemoteAddress(nullptr);
  }
```
So that ought to work....

tbkka replied:
I think I've addressed your concerns.  See below.

al45tair commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -987,6 +1042,60 @@ class MultiPayloadEnumTypeInfo: public EnumTypeInfo {
+      mask.andMask((uint64_t)0xFF00000000000007ULL, 0);
+    } else {
+      // Common to all 32-bit platforms
+      mask.andMask((uint32_t)0x00000003U, 0);
```
Ditto

tbkka replied:
Ditto.

tbkka commented on include/swift/Reflection/TypeRefBuilder.h:
```diff
@@ -729,6 +745,9 @@ class TypeRefBuilder {
   /// Get the unsubstituted capture types for a closure context.
   ClosureContextInfo getClosureContextInfo(RemoteRef<CaptureDescriptor> CD);
 
+  /// Get the multipayload enum projection informaion for a given TR
```
Done.

tbkka committed:
Typo

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -675,6 +711,12 @@ class BitMask {
   }
 
+  void makeZero() {
+    for (unsigned i = 0; i < size; ++i) {
```
Fixed.

tbkka committed:
Use memset

tbkka committed:
Make `andMask` consistent with usage

tbkka committed:
Do not break old, deprecated API

tbkka committed:
Some debugging utilities

tbkka committed:
Use a helper to produce the right type info

tbkka committed:
Be a little smarter about choosing a spare bit mask for pointers

 * If the target is the same size (32-bit or 64-bit) as the host, assume it's the same as the host
 * Otherwise, use ARM64 or ARM32 for 64-bit or 32-bit targets

Note: this should only get used for the fallback spare bits calculation when
compiler information is unavailable.  (Also, this is not the only place we
assume the target and host are the same architecture.)

tbkka committed:
Get the correct spare bit info from the type system

tbkka committed:
Use the correct spare bits mask; allow MPE record to grow in the future

The runtime wants the mask of "common spare bits for the
enum payloads", which we can generate inside the compiler
by AND-ing together the spare bits for each payload into
a single mask.

To allow the MPE record to resize in the future, express everything
as indexes in a "contents" array (of 32-bit values) with the size of
that array being part of the first element.  I've commented the
boilerplate for adding a new field (computing the index of the
added property, returning default value if the record isn't big enough).

With this, I'm finally passing my current tests, and have some local
validation that the compiler-provided spare bits mask in the
reflection record matches exactly the spare bits mask that the
runtime can compute from first principles.  This in turn gives
us some confidence that the runtime will be able to properly
reflect enum payloads for older binaries that lack the new reflection
record.

tbkka commented on include/swift/SwiftRemoteMirror/SwiftRemoteMirrorTypes.h:
```diff
@@ -91,6 +91,7 @@ typedef struct swift_reflection_mapping_info {
   swift_reflection_section_mapping_t capture;
   swift_reflection_section_mapping_t type_references;
   swift_reflection_section_mapping_t reflection_strings;
+  swift_reflection_section_mapping_t multipayload_enums; // TODO: Is this okay?
```
Ooops.  This needs to be removed.

tbkka replied:
Done.

tbkka commented on lib/IRGen/GenReflection.cpp:
```diff
@@ -1007,6 +1007,104 @@ void IRGenModule::emitBuiltinTypeMetadataRecord(CanType builtinType) {
   builder.emit();
 }
 
+ // TODO: This is copied from GenEnum.cpp; should we refactor to eliminate
```
TODO

tbkka replied:
@aschwaighofer - Copying this function from GenEnum.cpp was helpful.  Should I refactor to share one copy of this function?  Or do you prefer to keep utilities like this local to the source file they're used in?

aschwaighofer replied:
@tbkka I would make it available via `GenEnum.h`

tbkka replied:
Seems to be in GenValueWitness.cpp now; I exposed it via GenValueWitness.h so there's no longer a separate copy here.

tbkka committed:
Minor code simplification

tbkka committed:
Assert in the runtime that MPE reflection records are sane

tbkka committed:
Do not break legacy API

tbkka commented:
@swift-ci Please test

tbkka committed:
Update comment; remove unnecessary code

tbkka commented on include/swift/Reflection/Records.h:
```diff
@@ -374,6 +374,90 @@ class BuiltinTypeDescriptor {
+  }
+
+  void dump(std::ostream &stream) const {
+    // TODO
```
TODO: Implement dump()

tbkka replied:
Done.

tbkka committed:
Simply dump routine

tbkka committed:
Reduce the size of the reflection record

If `usePayloadSpareBits()` is not set, we don't need to store the bit mask size.

Break the bitmask size field into offset + size.  That allows us to omit leading zero bytes from the mask (we already omitted trailing zero bytes).

Add a new test reflect_Enum_MultiPayload_large.swift to verify that the offset handling works correctly.

tbkka committed:
Consolidate the two copies of getFormalTypeInContext()

tbkka commented:
@swift-ci Please test

tbkka committed:
Expand the failure messages here to help track down a bug

tbkka commented:
@swift-ci Please test Windows platform

tbkka committed:
Missing include

tbkka commented:
@swift-ci Please test Windows platform

tbkka committed:
More detail on this crash

tbkka commented:
@swift-ci Please test Windows platform

tbkka committed:
More detail on this crash

tbkka commented:
@swift-ci Please test Windows platform

tbkka committed:
Log the bytes and the type of data in the section

tbkka commented:
@swift-ci Please test Windows platform

tbkka committed:
print numbers, not characters

tbkka commented:
@swift-ci Please test Windows platform

tbkka committed:
Clarify a comment

tbkka commented:
@swift-ci Please test Linux platform

tbkka commented:
I added a bunch of `std::cerr <<` and now the Windows failure is dumping this information:
```
reflection section too small!
Section Type: MultiPayloadEnum
Remaining section size: 24(of 28), size of next record: 261124
Previous bytes: 20  0 10  0 
Next bytes:  7  0  0  0  0  0  0 ff 
```
That information makes no sense.  It's skipped one word so far, but MPE records are always at least two words.  And the second word is `0x00000007` which is also nonsense, since the MPE record code can't generate that particular value for the second word.

My only guess is that this is not really a MultiPayloadEnum section at all.  But staring at the source code isn't illuminating anything for me.

tbkka commented:
@compnerd - Any thoughts on this?  Another possibility that just occurred to me:  Maybe on Windows, we get some extra data at the front of these sections?  (This is probably the first reflection record type that's variably-sized, so it's almost conceivable that other reflection sections wouldn't be messed-up by extra bytes here.)

tbkka commented:
@swift-ci Please test macOS platform

tbkka committed:
Rename some variables for legibility

tbkka committed:
Try to make the index offsets easier to follow

tbkka committed:
Clean up the debug logging here a little bit

compnerd commented on include/swift/ABI/ObjectFile.h:
```diff
@@ -126,6 +131,8 @@ class SwiftObjectFileFormatCOFF : public SwiftObjectFileFormat {
     case protocs:
       return ".sw5prt$B";
+    case mpenum:
+      return ".sw5mpen";
```
I think that this here is the issue.  The `$B` is super critical.  It actually is used to coalesce the data between the begin and end makers.  This works with https://github.com/apple/swift/blob/main/stdlib/public/runtime/SwiftRT-COFF.cpp to ensure that the runtime is able to correctly locate the data.  You will also need to update the file to indicate a new metadata section.  There is a peer of it for ELF as well.

tbkka replied:
I took a guess and updated these files.  Does this look right now?

compnerd replied:
Yeah, that seems about right.

tbkka committed:
Add the new reflection section to the COFF and ELF declarations; declare the COFF section name with $B magic

tbkka commented:
@swift-ci Please test

tbkka committed:
Expose MPE Spare bit mask as a global var in target, fetch from host

This allows the host to query the target directly for the appropriate
spare bit mask info.  It requires the memory reader interface to
support `getSymbolAddress` but should otherwise be fully general.

Note: The growing stack of closures in TypeRefBuilder used to insulate this
non-template type from the template types it uses is getting pretty cumbersome.
Surely there's a better design?

tbkka commented:
@swift-ci Please test

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -1890,33 +2012,76 @@ class EnumTypeInfoBuilder {
+        // compilers.)
+        BitMask spareBitsMask(PayloadSize);
+        auto mpePointerSpareBits = TC.getBuilder().getMultiPayloadEnumPointerMask();
+        auto validSpareBitsMask = populateSpareBitsMask(Cases, spareBitsMask, mpePointerSpareBits);
```
@mikeash @al45tair Here's where I now ask the TypeRefBuilder to fetch the MPE pointer mask from the target and use that to compute the enum spare bit mask.

tbkka commented on include/swift/Reflection/TypeRefBuilder.h:
```diff
@@ -760,6 +839,40 @@ class TypeRefBuilder {
+  // If it can't read it, it will make an educated guess
+  // Note: This is a pointer-sized value stored in a uint64_t
+  // If the target is 32 bits, the mask is in the lower 32 bits
+  uint64_t getMultiPayloadEnumPointerMask() {
```
Here's the code in TypeRefBuilder to come up with the Enum pointer mask.  This first queries the target via `getSymbolAddress`.  But because tools like swift-reflection-dump don't implement `getSymbolAddress`, I need a fallback with logic similar to what I had before.

mikeash replied:
Looks good!

tbkka commented on include/swift/Reflection/TypeRefBuilder.h:
```diff
@@ -725,8 +775,37 @@ class TypeRefBuilder {
       OpaquePointerReader([&reader](remote::RemoteAddress address, unsigned size) -> llvm::Optional<remote::RemoteAbsolutePointer> {
         return reader.Reader->readPointer(address, size);
+      }),
+      OpaqueIntVariableReader(
```
This is the part I'm least happy with about this approach.  TypeRefBuilder has this pile of closures ostensibly to avoid making TypeRefBuilder itself a template, but it's getting increasingly unwieldy.  I'll take a stab at refactoring this a bit to make it cleaner, but any suggestions would be appreciated.  And of course, this "Int Variable Reader" compromises on the 32-bit/64-bit problem by requiring the caller to provide an explicit size, doing a 4-byte or 8-byte read as appropriate and then always returning the result in a `uint64_t`.

tbkka reviewed (commented):
I've addressed the concern @mikeash and @al45tair raised by filling in code to read the MPE spare bit mask for a pointer from the target.  It's more complex than I'd like but seems to work well enough.

tbkka commented:
@swift-ci Please test macOS Platform

tbkka committed:
Merge branch 'main' into tbkka-remoteMirror-MPE-spareBits-v2

tbkka committed:
Fix a merge conflict with the accessibility function section that's also being added

tbkka commented:
@swift-ci Please test

tbkka committed:
Merge branch 'main' into tbkka-remoteMirror-MPE-spareBits-v2

tbkka commented:
@swift-ci Please test

tbkka committed:
Merge branch 'main' into tbkka-remoteMirror-MPE-spareBits-v2

tbkka commented:
@swift-ci Please test

tbkka committed:
Enable debug assertions whenever !NDEBUG

tbkka committed:
Merge branch 'tbkka-remoteMirror-MPE-spareBits-v2' of github.com:tbkka/swift into tbkka-remoteMirror-MPE-spareBits-v2

tbkka commented:
@swift-ci Please test

tbkka committed:
New MPE test inspired by an LLDB test case

tbkka committed:
Disable mask computation for types containing enums

This means that if we don't have compiler information for a type that
has an enum in it somewhere, we'll refuse to project the enum value
because we don't know that we can do it correctly.

Also, disable the comparison of compiler-provided vs. locally-computed
masks for this case until this is fixed.

tbkka commented:
@swift-ci Please test

tbkka committed:
Merge remote-tracking branch 'github-apple/main' into tbkka-remoteMirror-MPE-spareBits-v2

tbkka commented:
@swift-ci Please test

tbkka commented:
The definitions for all reflection sections are now part of llvm-project.  Now that apple/llvm-project#3966 has landed, this should build and run cleanly.
